[Security Solution] expanded flyout - left section skeleton

PhilippeOberti opened the pull request:
## Summary

This PR leverages the expandable flyout work done in a previous [PR](https://github.com/elastic/kibana/pull/150240) and adds the alert details left section skeleton:
- add the expand/collapse details button in the flyout's header
- header with 4 tabs: Visualize, Insights, Investigations and History
- the last 3 tabs are empty but the first one (Visualize) has a button group with Session View and Analyser graph options (with placeholder components)

### How to test

- add `xpack.securitySolution.enableExperimental: ['securityFlyoutEnabled']` to the `kibana.json` file
- go to the Alerts page, and click on the expand detail button on any row of the table
- then click on the expand details button in the flyout's header

### Run tests and storybook
- `node scripts/storybook security_solution` to run Storybook
- `npm run test:jest --config ./x-pack/plugins/security_solution/public/flyout` to run the unit tests
- `yarn cypress:open-as-ci` but note that the integration/e2e tests have been written but are now skipped because the feature is protected behind a feature flag, disabled by default. To check them, add `'securityFlyoutEnabled'` [here](https://github.com/elastic/kibana/blob/main/x-pack/test/security_solution_cypress/config.ts#L50)

### Notes:

- the expandable flyout [work](https://github.com/elastic/kibana/pull/150240) needs to be merged first.
- the right section [work](https://github.com/elastic/kibana/pull/152047) also needs to be merged.

![Screenshot 2023-02-24 at 3 34 13 PM](https://user-images.githubusercontent.com/17276605/221297550-0e5bb3e8-daa4-4deb-ba26-e4d8b884a7ef.png)

https://github.com/elastic/security-team/issues/6098

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))

PhilippeOberti committed:
[Security Solution] expandable flyout left section skeleton

- expand/collapse button in right section header
- header with 4 tabs (visualize, insights, investigation and history)
- 3 empty tabs content and visualize has a button group for session view and analyser graph (empty components)

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

michaelolo24 commented on x-pack/plugins/security_solution/cypress/screens/document_expandable_flyout.ts:
```diff
@@ -0,0 +1,63 @@
+  FLYOUT_HEADER_TITLE_TEST_ID,
+} from '../../public/flyout/right/components/test_ids';
+
+export const DOCUMENT_DETAILS_FLYOUT_HEADER_TITLE = `[data-test-subj="${FLYOUT_HEADER_TITLE_TEST_ID}"]`;
```
As a thought for our cypress tests, rather than repeating all of the `[data-test-subj="blah"]`, why not just make a helper `const getTestSelector = (dataTestSubjectValue: string) => [data-test-subj="${dataTestSubjectValue}"]`? Might be useful for all of our selectors tbh

PhilippeOberti replied:
good idea, I extracted this in a function that I put under `/helpers/common.ts`, not sure if there is a better place...

michaelolo24 commented on x-pack/plugins/security_solution/public/flyout/left/components/analyze_graph.tsx:
```diff
@@ -0,0 +1,22 @@
+import { EuiText } from '@elastic/eui';
+import { ANALYSER_GRAPH_TEST_ID } from './test_ids';
+
+export const ANALYZE_GRAPH_ID = 'analyze_graph';
```
A nit, but I would unify on `analyze` vs `analyse` (which is what I think is used in the cypress tests)

michaelolo24 replied:
I was trying to figure out what was up...and TIL 😂 

<img width="750" alt="image" src="https://user-images.githubusercontent.com/17211684/223550839-ae7a7cc5-c770-4f64-9832-84123d33f4e3.png">

michaelolo24 replied:
😂 I actually kinda like analyse. Feels fancier 🍷 lol . Thanks for unifying it!

michaelolo24 commented on x-pack/plugins/security_solution/public/flyout/left/context.tsx:
```diff
@@ -0,0 +1,48 @@
+} & Partial<LeftPanelProps['params']>;
+
+export const LeftPanelProvider = ({ id, indexName, children }: LeftPanelProviderProps) => {
+  const contextValue = {
```
I would wrap these context objects in a `useMemo` as any time a parent re-renders a new object is going to be created, even if the values stay the same. Alternatively, you could wrap the entire provider in a `React.memo`

michaelolo24 replied:
Also, since the context can be undefined...maybe something like this to avoid the casting?

```
  const contextValue = useMemo(
    () =>
      id && indexName
        ? {
            eventId: id,
            indexName,
          }
        : undefined,
    [id, indexName]
  );
  ```

PhilippeOberti replied:
yup I like this a lot better, thanks!

michaelolo24 commented on x-pack/plugins/security_solution/public/flyout/left/tabs.tsx:
```diff
@@ -0,0 +1,54 @@
+  content: React.ReactElement;
+}>;
+
+export const tabs: LeftPanelTabsType = [
```
Just making a note that in the future, depending on how GA-ing this goes, we may want an `enabled` flag here to track tabs we may want to enable / leave disabled for a future release

michaelolo24 replied:
Don't need to add this now by any means, just something to keep in mind

michaelolo24 commented on x-pack/plugins/security_solution/public/flyout/left/tabs/test_ids.ts:
```diff
@@ -0,0 +1,18 @@
+ * 2.0.
+ */
+
+export const VISUALIZE_TAB_BUTTON_GROUP_TEST_ID =
```
Thank you for adding all these test ids! 🙇🏾

michaelolo24 commented on x-pack/plugins/security_solution/public/flyout/right/components/expand_detail_button.stories.tsx:
```diff
@@ -0,0 +1,54 @@
+  title: 'Flyout/ExpandDetailButton',
+};
+
+export const Expand: Story<void> = () => {
```
This is sweet. You should demo this during the next team wide demo 😄

PhilippeOberti replied:
yeah I really like Storybook, it's nice when you review a PR to be able to look at a component and its different states. Happy to demo it!

PhilippeOberti committed:
fix PR comments

- analyse => analyze
- use useMemo in contexts
- create cypress selectors in reusable functions

michaelolo24 commented on x-pack/plugins/security_solution/cypress/helpers/common.ts:
```diff
@@ -0,0 +1,19 @@
+ * Helper function to generate selector by data-test-subj
+ * @param dataTestSubjectValue the value passed to the data-test-subj property of the DOM element
+ */
+export const getDataTestSubjectSelector = (dataTestSubjectValue: string) =>
```
Thanks for adding! It would be cool in the future (not for this PR, but for one that revamps our cypress tests) where we add custom commands for our test suite: a la:

https://docs.cypress.io/guides/references/best-practices#Real-World-Example

michaelolo24 reviewed (approved):
Thanks for making these changes, this is awesome!!
